src/test_embedding.py

```diff
@@ -0,0 +1,278 @@
+"""
+Hydra-based Embedding Extractor (Direct JSON)
+=============================================
+This version skips the HDF5/DataModule pipeline and directly
+builds graph features from JSON discussion trees for embedding extraction.
+
+Usage:
+------
+python src/test_embedding.py \
+  --json src/test_data.json \
+  --ckpt src/logs/pretrain/runs/2025-10-23_22-41-28/checkpoints/last.ckpt \
+  --device cuda \
+
+"""
+
+# the batch size command is removed as it actually valids the parameters,
+# need to see why
+
+import os
+import sys
+import json
+import torch
+import argparse
+import numpy as np
+from omegaconf import OmegaConf
+from hydra.utils import instantiate
+from transformers import AutoTokenizer, AutoImageProcessor
+from PIL import Image
+from data.types import GraphFeatures, TextFeatures, ImageFeatures
+from data import collator_utils
+from tasks import dataset_utils as dut
+
+# ---- Pathing ----
+ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+sys.path.extend([ROOT, os.path.join(ROOT, "src")])
+
+
+#  Convert a tree jsn to feature dicts, the graph
+#  watch: Similar to data.dataset_utils.tree_to_graph_features but simplified
+def tree_to_graph_features(tree, text_tokenizer, image_processor, root_dir="."):
+    """Flatten a discussion tree into tensors for model input."""
+    dut.compute_relative_distance(tree)
+
+    flat = {
+        "id": [],
+        "parent_id": [],
+        "text": [],
+        "images": [],
+        "rotary_position": [],
+    }
+
+    def traverse(node, parent_id=None):
+        if parent_id is None:
+            parent_id = node["id"]
+        flat["id"].append(node["id"])
+        flat["parent_id"].append(parent_id)
+        flat["text"].append(dut.clean_text(node.get("body", "")))
+        flat["images"].append(
+            node.get("images", [None])[0] if node.get("images") else None
```

Comment:
this can be simplied to just `node.get("images")` (https://docs.python.org/3.8/library/stdtypes.html#dict.get) 
Calling `node.get("images")` will return `None` if the key is not present.

Reply:
On second thought, it seems like the fact that `images` being a list complicates things, ugh. either way,  `node.get("images", [None])[0]` will work too.